nni/compression/experiment/config/vessel.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+import base64
+import io
+from dataclasses import dataclass, asdict
+from pathlib import Path
+from typing import Any, Callable, Tuple, overload
+
+import torch
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from nni.algorithms.compression.v2.pytorch.utils.constructor_helper import OptimizerConstructHelper
+from nni.common import dump, load
+from nni.experiment.config.base import ConfigBase
+
+
+@dataclass(init=False)
+class CompressionVessel(ConfigBase):
+    """
+    This is an internal class that helps serialize model-related parameters during model compression.
+
+    # FIXME: In fact, it is not a `Config`, the only reason it is a `Config` right now is that its data attribute
+    # will go into the search space as a single choice field. Need to refactor after the experiment config is stable.
+    """
+
+    model: str
+    finetuner: str
+    evaluator: str
+    dummy_input: str
+    trainer: str | None
+    optimizer_helper: str | None
+    criterion: str | None
+    device: str
+
+    @overload
+    def __init__(self, model: str, finetuner: str, evaluator: str, dummy_input: str,
+                 trainer: str, optimizer_helper: str, criterion: str, device: str):
+        ...
+
+    @overload
+    def __init__(self, model: Module, finetuner: Callable[[Module], None], evaluator: Callable[[Module], float],
+                 dummy_input: Tensor, trainer: Callable[[Module, Optimizer, Callable[[Any, Any], Any]], None] | None,
+                 optimizer_helper: Optimizer | OptimizerConstructHelper | None,
+                 criterion: Callable[[Any, Any], Any] | None, device: str | torch.device):
+        ...
```

Comment:
why overload here?

Reply:
because we want to pass in either unserialized instances or strings [serialized instances].

In other words, as a user interface, we want users to only consider the case of non-serialized instances. The case of passing in serialized strings is an advanced usage and should be used after at least familiarity with nni's serialization.